r2i/onnxrt/engine.cc

```diff
@@ -0,0 +1,281 @@
+ * a software license from RidgeRun.  All source code changes must be provided
+ * back to RidgeRun without any encumbrance.
+ */
+
+#include "r2i/onnxrt/engine.h"
+
+#include <core/common/exceptions.h>
+#include <core/session/onnxruntime_cxx_api.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "r2i/onnxrt/frame.h"
+#include "r2i/onnxrt/model.h"
+#include "r2i/onnxrt/prediction.h"
+
+namespace r2i {
+namespace onnxrt {
+
+Engine::Engine () : state(State::STOPPED), model(nullptr) {
+}
+
+Engine::~Engine () {
+  this->Stop();
+}
+
+RuntimeError Engine::SetModel (std::shared_ptr<r2i::IModel> in_model) {
+  RuntimeError error;
+
+  if (State::STOPPED != this->state) {
+    error.Set (RuntimeError::Code::WRONG_ENGINE_STATE,
+               "Stopped model before setting a new state");
+    return error;
+  }
+
+  if (nullptr == in_model) {
+    error.Set (RuntimeError::Code:: NULL_PARAMETER,
+               "Received null model pointer");
+    return error;
+  }
+  auto model = std::dynamic_pointer_cast<r2i::onnxrt::Model, r2i::IModel>
+               (in_model);
+
+  if (nullptr == model) {
+    error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
+               "The provided model is not an ONNXRT model");
+    return error;
+  }
+
+  if (nullptr != this->model) {
+    this->model = nullptr;
+  }
+
+  this->model = model;
+
+  return error;
+}
+
+RuntimeError Engine::Start ()  {
+  RuntimeError error;
+
+  if (State::STARTED == this->state) {
+    error.Set (RuntimeError::Code::WRONG_ENGINE_STATE,
+               "Engine already started");
+    return error;
+  }
+
+  if (nullptr == this->model) {
+    error.Set (RuntimeError::Code:: NULL_PARAMETER,
+               "Model not set yet");
+    return error;
+  }
+
+  this->state = State::STARTED;
+
+  return error;
+}
+
+RuntimeError Engine::Stop () {
+  RuntimeError error;
+
+  if (State::STOPPED == this->state) {
+    error.Set (RuntimeError::Code::WRONG_ENGINE_STATE,
+               "Engine already stopped");
+  }
+
+  this->state = State::STOPPED;
+
+  return error;
+}
+
+std::shared_ptr<r2i::IPrediction> Engine::Predict (std::shared_ptr<r2i::IFrame>
+    in_frame, r2i::RuntimeError &error) {
+  ImageFormat frame_format;
+  int frame_width = 0;
+  int frame_height = 0;
+  int frame_channels = 0;
+  std::vector<Ort::Value> output_tensor;
+  std::vector<int64_t> input_node_dims;
+  size_t input_image_size = 0;
+
+  auto prediction = std::make_shared<Prediction>();
+
+  error.Clean ();
+
+  if (State::STARTED != this->state) {
+    error.Set (RuntimeError::Code::WRONG_ENGINE_STATE,
+               "Engine not started");
+    return nullptr;
+  }
+
+  auto frame = std::dynamic_pointer_cast<Frame, IFrame> (in_frame);
+  if (!frame) {
+    error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
+               "The provided frame is not an onnxrt frame");
+    return nullptr;
+  }
+
+  std::shared_ptr<Ort::Session> session = this->model->GetOnnxrtSession();
```

Comment:
This will attempt to delete the session using `delete`. I don't think this is right.